src/dune_pkg/repository_id.ml

```diff
@@ -0,0 +1,83 @@
+open Stdune
+module Process = Dune_engine.Process
+module Display = Dune_engine.Display
+
+type t =
+  | Git_hash of string
+  | Opam_stamp of string
+
+let equal a b =
+  match a, b with
+  | Git_hash a, Git_hash b | Opam_stamp a, Opam_stamp b -> String.equal a b
+  | _, _ -> false
+;;
+
+let encode : t -> Dune_sexp.t = function
+  | Git_hash commitish ->
+    List
+      [ Dune_lang.atom_or_quoted_string "Git_hash"
+      ; Dune_lang.atom_or_quoted_string commitish
+      ]
+  | Opam_stamp stamp ->
+    List
+      [ Dune_lang.atom_or_quoted_string "Opam_stamp"
+      ; Dune_lang.atom_or_quoted_string stamp
+      ]
+;;
+
+let decode =
+  let open Dune_sexp.Decoder in
+  let+ constr, stamp = pair string string in
+  match constr with
+  | "Git_hash" -> Git_hash stamp
+  | "Opam_stamp" -> Opam_stamp stamp
+  | _ -> failwith "TODO: parsing failure"
+;;
+
+let to_dyn = function
+  | Git_hash commitish -> Dyn.Variant ("Git_hash", [ String commitish ])
+  | Opam_stamp stamp -> Dyn.Variant ("Opam_stamp", [ String stamp ])
+;;
+
+let repo_id_of_stamp dir =
+  let local = Path.Local.of_string "repo" in
+  let repo_file = Path.append_local dir local in
+  match Path.exists repo_file with
+  | false ->
+    (* no repo file, we can't read this, bail *)
+    Fiber.return None
+  | true ->
+    (match OpamParser.FullPos.file (Path.to_string repo_file) with
+     | exception Stdlib.Parsing.Parse_error -> Fiber.return None
+     | opam_file ->
+       let stamp =
+         List.find_map opam_file.file_contents ~f:(fun { pelem; _ } ->
+           match pelem with
+           | Variable ({ pelem = "stamp"; _ }, { pelem = String stamp; _ }) -> Some stamp
+           | _ -> None)
+       in
+       stamp |> Option.map ~f:(fun s -> Opam_stamp s) |> Fiber.return)
+;;
+
+let of_path dir =
+  let open Fiber.O in
+  let failure_mode = Process.Failure_mode.Return in
+  let display = Display.Silent in
+  let path = Env.get Env.initial "PATH" |> Option.value_exn |> Bin.parse_path in
+  match Path.is_directory dir with
```

Comment:
This function swallows permission errors. We should avoid using it and use `Path.stat` when possible